minigui/README.md

```diff
@@ -53,36 +69,247 @@ A UI for Minigo
+```
+
+#### Update messages
+
+While an engine is thinking (in response to a `genmove` command or pondering
+while it's the other player's turn), an engine should also send incremental
+updates for the current position, containing variations considered, expected win
+rate, number of visits, etc.
+
+These updates should be printed as a JSON object (again on a single line) with
+the prefix `mg-update:`.
+
+ - `id: string`: the position's id.
+ - `n?: number`: the number of reads for the position.
+ - `q?: number`: estimated winrate for the position from black's perspective in
+   the range [-1, 1].
+ - `pv?: string`: the GTP coordinate of the variation in `variations` that
+   corresponds to the principal variation.
+ - `treeStats?: TreeStats`: stats for the current search tree (see below).
+ - `variations?: {string: Variation}`: a map from GTP coordinate to a variation
+   (see below). Some Minigui modes also display a live view of the current
+   variation being searched. This should be written to `variations` under the
+   `"live"`.
+
+Where `TreeStats` is:
+ - `numNodes: number`: total number of nodes in the search tree.
+ - `numLeafNodes: number`: number of leaf nodes in the search tree.
+ - `maxDepth: number`: maximum depth of the search tree.
+
+And `Variation` is:
+ - `n: number`: number of reads for this variation.
+ - `q: number`: expected winrate for the variation from black's perspective in
+   the range [-1, 1].
+ - `moves: string[]`: the list of moves in this variation as GTP coordinates.
+   Note that the first coordinate is present in the `move` list even though it
+   is also used as the key in the parent `variations` object.
+
+We recommend only reporting variations that your engine considers interesting.
+
+Here's a simple example from Minigo, again pretty-printed to aid legibility:
+
+```
+mg-update: {
+  "id": "0x55834bd840f0",
+  "n": 23,
+  "q": -0.9926629662513733,
+  "variations": {
+    "B14": {
+      "moves": ["B14"],
+      "n": 1,
+      "q": -0.9774199724197388
+    },
+    "C14": {
+      "moves": ["C14","A13"],
+      "n": 2,
+      "q":-0.9983565211296082
+    },
+    "D15": {
+      "moves": ["D15","A13"],
+      "n": 2,
+      "q": -0.9999065399169922
+    },
+    "H8":{
+      "moves": ["H8","G10"],
+      "n": 3,
+      "q": -0.9687618017196655
+    },
+    "J4": {
+      "moves": ["J4","J5","K5"],
+      "n": 4,
+      "q": -0.9998496770858765
+    },
+  }
+}
+```
+
+### Synchronizing stdout and stderr
+
+All `mg-position` and `mg-update` generated in response to a GTP command must be
+printed and flushed to `stderr` because GTP mandates what can and cannot be
+written to `stdout`.
+
+When processing a GTP command, the engine must ensure that both `stdout` and
+`stderr` stay in sync by printing and flushing the string `__GTP_CMD_DONE__` to
+`stderr` before it writes the GTP response to `stdout`.
```

Comment:
can you give an example of this.

e.g.

genmove
stderr: aoeu
stderr: aoeu
stderr: __GTP_CMD_DONE__
stdout: D4